core/modfunc.go

```diff
@@ -361,6 +382,69 @@ func (fn *ModuleFunction) Call(ctx context.Context, opts *CallOpts) (t dagql.Typ
 	return returnValueTyped, nil
 }
 
+func extractError(ctx context.Context, client *buildkit.Client, baseErr error) (dagql.ID[*Error], bool, error) {
```

Comment:
Not a blocker because this current approach works, but I do feel like we could skip all the shenanigans of writing the error back to the module container and then reading it from there. The reason we currently do that for the (non-error) return value was to be setup for caching function calls (since the output would be cached in the filesystem of the module container and thus integrated w/ buildkit's caching), but for error cases none of that should matter.

I think `ReturnError` could just store the error in the client state, where it can be retrieved by anything that needs it.

Again, not a blocker, just seems like it would probably be simpler and delete a few chunks of code.

TBH if we also stored the actual return value in the client state too so we could skip reading from the filesystem I wouldn't be shocked if we saved a just-barely-noticeable amount of overhead to function calls, though that's getting way out of scope, just musing...

Reply:
That makes sense - OK to fix this up after? Leaning towards merging sooner to give it more time to "cook."